### DISCUSSION TITLE
com_ajax

### DISCUSSION THREAD

**Discussion by fvlasenko** [joomla/joomla-cms#47157] [2. Cold Features]
### What needs to be fixed
com_ajax

### Why this should be fixed
Add functional ajax in admin module

### How would you fix it
replace 
`$helperFile = JPATH_BASE . '/modules/mod_' . $module . '/helper.php';`
to
`$basePath   = ($table->client_id) ? JPATH_ADMINISTRATOR : JPATH_SITE;
$helperFile = $basePath . '/modules/mod_' . $module . '/helper.php';`
remove
`// Load language file for module
$basePath = JPATH_BASE;`
### Side Effects expected
No

**Comment by Fedik** [comment]
`JPATH_BASE` already holds a correct path, for an administrator and for a site.
If you need access to administrator module then do request to `/administrator`.

**Comment by SharkyKZ** [comment]
Duplicate https://github.com/joomla/joomla-cms/issues/25684.

Using client ID would be fine except that we can't determine it based on module name. Modules with the same name can be installed to both backend and frontend, e.g. like core `mod_feed` or `mod_custom` modules.

**Comment by Fedik** [comment]
I do not think it is an issue, you just have to do a correct request:

for site modules: `/index.php?option=com_ajax&module=foobar&format=json`
for admin modules: `/administrator/index.php?option=com_ajax&module=foobar&format=json`

Same as with any component.

**Comment by SharkyKZ** [comment]
This is no good if you want to use Ajax in module form.

**Comment by Fedik** [comment]
why you ever need that? :)

It the same as for components, you cannot execute Site component by doing request to Backend, and vice versa. The same applies to the modules.

**Comment by SharkyKZ** [comment]
To update form fields using AJAX. It's a reasonable use case.

**Comment by Fedik** [comment]
Ah okay, I finally get it, it for a module configuration form, not for a module itself :smile:  

Well, then,  just add `&client=0` to query, and update com_ajax to use it.
So you can call site module from backend. Eg:
`/administrator/index.php?option=com_ajax&module=foobar&format=json&client=0`

But, it should be forbidden to call administrator module from the site. So if someone call `&client=1` from the site then  com_ajax should throw an error.

**Comment by fvlasenko** [comment]
@Fedik i use this in admin panel modules
client="administrator"
